### PR TITLE
Support `make_array` null handling in nested version

### LINKED ISSUE #7136: `make_array` null handling with nested version
### Is your feature request related to a problem or challenge?

Follow up on #6900. Implement array() nested version.
It seems calling array() for list return flattened list according to the comment

https://github.com/apache/arrow-datafusion/blob/2cf5f5b5bb824598de185d64c541c52c930728cf/datafusion/physical-expr/src/array_expressions.rs#L243-L254

However, I think make_array should wrap the elements into array, the expected behavior 
`make_array([A,X], []) -> [[A,X], []]`,
`make_array([null, Y], [Q,R,S]) -> [[null,Y],[Q,R,S]]` and 
`make_array([C,Z], null) -> [[C,Z], null]`

Main branch:
```
query ?
select make_array(['A','X'], []);
----
[[A, X], ]

query ?
select make_array(make_array(null,'Y'), make_array('Q','R','S'));
----
[[, Y], [Q, R, S]]

query error DataFusion error: Optimizer rule 'simplify_expressions' failed
caused by
Internal error: failed to downcast\. This was likely caused by a bug in DataFusion's code and we would welcome that you file an bug report in our issue tracker
select make_array(['C','Z'], null);
```

As a result, only the third case fails, the first case is acceptable but should we expected `[[A, X], []]` to differentiate with `null` can also be discussed

### Describe the solution you'd like

Fix the third case and update the comments.
Optional: Expect the first case to return `[[A, X], []]`

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by jayzhan211**
## Which issue does this PR close?

Closes #7136.

### CODE REVIEW

**Comment by jayzhan211** [comment]
@izveigor @alamb Ready for review.

File: datafusion/physical-expr/src/array_expressions.rs [hunk 1]
```diff
@@ -327,21 +363,34 @@ fn array(values: &[ColumnarValue]) -> Result<ColumnarValue> {
...
     }
 
     match data_type {
-        DataType::Null => Ok(ColumnarValue::Scalar(ScalarValue::new_list(
+        None => Ok(ColumnarValue::Scalar(ScalarValue::new_list(
             Some(vec![]),
             DataType::Null,
         ))),
-        _ => Ok(ColumnarValue::Array(array_array(
+        Some(DataType::Null) => {
```
[the remark this answers is not in the document]
Comment (jayzhan211) [reply datafusion/physical-expr/src/array_expressions.rs:382]: @izveigor , duckdb consider make_array(array[]) and make_array(array[null]) difference things, but in datafusion we get `Scalar(List([NULL]))` in `make_scalar_function_with_hints` args. Do you know where can we change the args we get before create_physical_expr?

**Review by alamb — APPROVED** [review]
Looks like an improvement to me -- thank you @jayzhan211 

Any concerns about merging this @izveigor ?

**Comment by alamb** [comment]
Thanks @izveigor  and @jayzhan211 !

**Pull request merged**